### PR TITLE
docs(adr): ADR-016: Reactions module

### PR DESCRIPTION
**Pull request by RiccardoM**
## Description

This PR contains the ADR for the new `x/reactions` module allowing users to react to posts and subspace owners to register custom reaction within their own subspaces.

References: #890

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://docs.cosmos.network/v0.44/building-modules/intro.html)
- [ ] included the necessary unit and integration [tests](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: docs/architecture/adr-016-reactions-module.md [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+message MsgRemoveRegisteredReactionResponse {}
+
+// MsgSetReactionsParams represents the message to be used when setting 
+// a subspace reactions params
+message MsgSetReactionsParams {
+  // Id of the subspace for which to set the params
+  uint64 subspace_id = 1;
+
+  // Params related to RegisteredReactionValue reactions
+  RegisteredReactionValueParams registered = 2;
```
Comment (bragaz) [inline docs/architecture/adr-016-reactions-module.md:240]: I think you miswrote here
```suggestion
  RegisteredReactionValueParams reactions = 2;
```

**Review by bragaz — COMMENTED** [review]
To me everything looks good and pretty straightforward and easy customisable by the devs.
One question only:
Since in the original implementation that we have in the `backlog` branch we were checking the validity of emojis with a library, I assume here we will NOT check it, and let third parties devs takes care of it, right?
I think shifting that to them will make easier for us to maintain the module since it will have less dependencies to rely on.

File: docs/architecture/adr-016-reactions-module.md [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+// Query defines the gRPC querier service.
+service Query {
+  // Reactions allows to query the reactions present inside a subspace
+  rpc Reactions(QueryReactionsRequest) returns QueryReactionsResponse {
+    option (google.api.http).get = "/desmos/reactions/v1/{subspace_id}/reactions";
+  }
+
+  // RegisteredReactions allows to query the registered reaction of a subspace
+  rpc RegisteredReactions(QueryRegisteredReactionsRequest) returns QueryRegisteredReactionsResponse {
+    option (google.api.http).get = "/desmos/reactions/v1/{subspace_id}/registered_reactions";
```
Comment (dadamu) [inline docs/architecture/adr-016-reactions-module.md:264]:
```suggestion
    option (google.api.http).get = "/desmos/reactions/v1/{subspace_id}/registered-reactions";
```

File: docs/architecture/adr-016-reactions-module.md [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+  string display_value = 3;
+  
+  // User adding the supported reaction
+  string user = 4;
+}
+
+// MsgAddRegisteredReactionResponse represents the 
+// Msg/AddRegisteredReaction response type
+message MsgAddRegisteredReactionResponse {
+  // If of the newly registered reaction
```
Comment (dadamu) [inline docs/architecture/adr-016-reactions-module.md:212]:
```suggestion
  // Id of the newly registered reaction
```

File: docs/architecture/adr-016-reactions-module.md (hunk 1 above)
Comment (RiccardoM) [reply docs/architecture/adr-016-reactions-module.md]: I've updated the name to make it clearer

**Comment by RiccardoM** [comment]
> Since in the original implementation that we have in the `backlog` branch we were checking the validity of emojis with a library, I assume here we will NOT check it, and let third parties devs takes care of it, right? I think shifting that to them will make easier for us to maintain the module since it will have less dependencies to rely on.

Yes, that's right. We will leave the subspace owners/moderators complete freedom to decide how the reactions users can use should be validated. To do so, I've also now added a new field inside the `FreeTextValueParams` that allows subspace owners to set a custom regex to be used to check the validity of `FreeTextValue` reactions which should provide even more customization opportunities

**Review by dadamu — APPROVED** [review]
All looks good to me.

File: docs/architecture/adr-016-reactions-module.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+```protobuf
+syntax = "proto3";
+
+// SubspaceReactionsParams contains the params related to a single subspace reactions
+message SubspaceReactionsParams {
+  // Id of the subspace for which these params are valid
+  uint64 subspace_id = 1;
+  
+  // Params related to RegisteredReactionValue reactions
+  RegisteredReactionValueParams registered = 2;
```
Comment (RiccardoM) [inline docs/architecture/adr-016-reactions-module.md:105]:
```suggestion
  RegisteredReactionValueParams registered_reactions = 2;
```

File: docs/architecture/adr-016-reactions-module.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+```protobuf
+syntax = "proto3";
+
+// SubspaceReactionsParams contains the params related to a single subspace reactions
+message SubspaceReactionsParams {
+  // Id of the subspace for which these params are valid
+  uint64 subspace_id = 1;
+  
+  // Params related to RegisteredReactionValue reactions
+  RegisteredReactionValueParams registered_reactions = 2;
```
Comment (RiccardoM) [inline docs/architecture/adr-016-reactions-module.md:105]:
```suggestion
  RegisteredReactionValueParams registered_reaction = 2;
```

**Pull request merged**